[collectd 6, WIP] gpu_sysman.c: use OTEL style metric names + add .unit types

eero-t opened the pull request:
ChangeLog: gpu_sysman.c: use OTEL style metric names + add .unit types

This adds unit member to all Sysman plugin metrics, and changes their names to match OTEL spec better.

Test code converts ".unit" names to previously used unit names (+ adds "_total" suffix for counters), to minimize changes to rest of the test code.

Before merging:
* Prometheus plugin needs generic support for converting `<unit>/<unit>` to `units_per_unit` (Sysman plugin uses `By/s` units for few metrics)
* I'll change Sysman plugin metrics with `us` & `uj` units to use base units instead, after https://github.com/collectd/collectd/pull/4266 is merged (Prometheus plugin lacks `uj` support)
* Still undecided on what to do with `MHz` units, add it to Prometheus plugin, or switch to `Hz`?
* After Prometheus side supports the needed units, I'll need to update few internal tests and try this out...

octo commented on src/gpu_sysman_test.c:
```diff
@@ -755,40 +755,75 @@ static int validate_and_reset_saved_metrics(unsigned int base_rounds,
+    const char *unit;
+    const char *name;
+  } units[] = {
+      {"1", "_ratio"},     {"By", "_bytes"}, {"By/s", "_bytes_per_second"},
+      {"Cel", "_celsius"}, {"{error}", ""},  {"MHz", "_mhz"},
+      {"uj", "_ujoules"},  {"us", "_usecs"}, {"W", "_watts"},
```
That the code needs to add a suffix based on the unit is an indicator that the metric names are not sufficiently unique. Note that most systems used to aggregate metrics will not be able to filter/group by the unit.

eero-t replied:
Thanks, that's good to know!   I'll drop that part of the tester if all metrics get satisfactory unique names.

octo reviewed (changes requested):
Three main points:

* All the error counters (minus the "all" one) should be collated into one metric.
* Many metrics of different types (e.g. temperature, ratio) have the same name. These need to be split into seperate metrics.
* We should align this plugin with the [semantic convention for GPUs](https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hwgpu---gpu-metrics).

eero-t commented on src/gpu_sysman.c:
```diff
@@ -59,7 +59,7 @@
 
 #define PLUGIN_NAME "gpu_sysman"
-#define METRIC_PREFIX "collectd_" PLUGIN_NAME "_"
+#define METRIC_PREFIX "sysman.gpu."
```
Sounds reasonably, but have you suggestion for differentiating GPU plugins metrics from from each other? Adding vendor labels to all metrics would be very wasteful...

How / could I add vendor info as resource (i.e. output only once per query, not per-metric)?

eero-t replied:
Thanks!

I guess I need to generate per-GPU label sets at plugin init time, add them to families in the common submit function, and reset those label sets only when plugin is uninitialized.

Btw. There are already few per-GPU labels added to all metric families in the submit function (`pci_dev`, `pci_dev`, `dev_file`) which I can move over to resource labels, see: https://github.com/collectd/collectd/blob/collectd-6.0/src/gpu_sysman.c#L770

eero-t replied:
@octo Storing per-GPU info in per-metric-family resource, instead of metric labels, breaks metrics reporting.

When multiple GPUs are reporting (similar) metric values, so that only difference is in their resource label values, `write_prometheus` will report metrics and target_info only for the _first_ GPU, and _ignore metrics submitted for all the other devices_.

PS. logging metrics directly from the plugin itself was really useful for debugging this.

eero-t replied:
If I add unique per-GPU label (back) to metrics, then  `write_prometheus` (again) reports metrics for all GPUs.

However, it still reports `target_info` only for the first GPU.  I.e. resource handling is not really useful in its current form. :-/

eero-t replied:
With #4284, and the commit adding GPU ID as "service.instance.id" this started working OK.   I.e. I can next add rest of the labels suggested by the spec.

eero-t commented on src/gpu_sysman.c:
```diff
@@ -900,54 +901,54 @@ static bool gpu_ras(gpu_device_t *gpu) {
       case ZES_RAS_ERROR_CAT_RESET:
         help = "Total count of HW accelerator resets attempted by the driver";
-        catname = METRIC_PREFIX "resets_total";
+        catname = METRIC_PREFIX "errors.resets";
```
> Can the errors.* metrics be mapped to hw.errors, with the `hw.error.type` label indicating the type of error?

Errors part of spec looks really annoying, and seems to require something stupid like this (instead of what you proposed):
`hw.errors{hw.type="gpu",hw.error.type="uncorrectable",hw.error.gpu.type="reset"}`

Whereas I'd rather use something (clearly) more readable, and more consistent with the other GPU metrics in the spec:
`hw.gpu.errors.resets{hw.error.type="uncorrectable"}`

Comments?

eero-t replied:
Btw. spec already has `hw.gpu.power`: https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hwgpu---gpu-metrics

and `hw.power{hw.type="gpu"}`: https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hw---common-hardware-metrics

So same for errors seems natural.

octo replied:
My guess is that the idea is to use `hw.errors` (and `hw.status`) for all hardware, so that you can combine them in one graph, showing you any hardware error that occurred on the system. The `hw.type` label allows you to aggregate by hardware type, e.g. filter for GPU errors or sum up GPU errors.

Judging from other sections, data points are supposed to have one other label, `hw.error.type`. The values ("corrected", "uncorrected") are only examples. This is where the differentiation between different errors should happen. For the gpu_sysman plugin, we should set this to "reset", "programming", "driver", …

The duplication between `hw.power`, `hw.gpu.power` is weird. I'd probably prefer `hw.power` so users can aggregate the power consumption of different components.

octo replied:
> > The values ("corrected", "uncorrected") are only examples.
> 
> No, they're something that need to be in GPU metrics / were already provided by Sysman metrics `type` label (which needs now `hw.error.` prefix). I.e. the actual error names like `reset` etc, would need to use some other label, maybe `hw.error.name`?

Other devices suggest other values for `hw.error.type`, e.g. network interfaces suggest "zero_buffer_credit, crc, etc." and disks suggest "bad_sector, write, etc.". For me it is quite clear that this is meant as "whatever is relevant for the device goes here" kind of placeholders.

If all GPU errors come in corrected and uncorrected variants, I'd actually suggest to introduce a new label, `hw.error.correction` and set this to either "corrected" or "uncorrected" instead.

> As to power aggregation, there's separate metric category for whole host power

That makes sense to me. The host power is an aggregate of sorts, so it makes sense to define a separate metric for it.

> Weirdly it does not list actual power usage, just limit & utilization.

That's really weird.

>  And worse, it does not require label for which component's PSU the data is for (whole host, GPU, some other component?), WTF?

They do recommend a `serial_number` attribute, which can use used to differentiate different power supplies. I think if there are dedicated power supplies for mainboard/cpu on one hand and GPU(s) on the other hand, then it would make perfect sense to me to tag them accordingly.

eero-t replied:
> If all GPU errors come in corrected and uncorrected variants, I'd actually suggest to introduce a new label, hw.error.correction and set this to either "corrected" or "uncorrected" instead.

Some of them are not correctable (e.g. resets), but most are: https://spec.oneapi.io/level-zero/latest/sysman/PROG.html#querying-ras-errors

Both above oneAPI Level Zero spec and OpenTelemetry spec state that to be error "type", so I would use that.

As to whether the error is GPU reset, in HW programming, in cache or other location, the term used by the L0 spec for that is "category": https://spec.oneapi.io/level-zero/latest/sysman/api.html#zes-ras-error-type-t

So, maybe the label could be "hw.error.category"?

> They do recommend a serial_number attribute, which can use used to differentiate different power supplies.

I do not see how user is supposed to determinate one "random" (serial) number to mean GPU, and another "random" (serial) number to mean PSU. :-)

eero-t commented on src/gpu_sysman.c:
```diff
@@ -1259,20 +1262,24 @@ static bool gpu_mems_bw(gpu_device_t *gpu) {
+  char *name = METRIC_PREFIX "memory.bw";
   metric_family_t fam_ratio = {
       .help = "Average memory bandwidth usage ratio (0-1) over query interval",
-      .name = METRIC_PREFIX "memory_bw_ratio",
       .type = METRIC_TYPE_GAUGE,
+      .name = name,
+      .unit = "1",
   };
   metric_family_t fam_rate = {
       .help = "Memory bandwidth usage rate (in bytes per second)",
-      .name = METRIC_PREFIX "memory_bw_bytes_per_second",
       .type = METRIC_TYPE_GAUGE,
+      .name = name,
+      .unit = "By/s",
   };
   metric_family_t fam_counter = {
       .help = "Memory bandwidth usage total (in bytes)",
-      .name = METRIC_PREFIX "memory_bw_bytes_total",
       .type = METRIC_TYPE_COUNTER,
+      .name = name,
+      .unit = "By",
   };
```
I think OTEL naming requirements / recommendations are badly thought out and inconsistent (e.g. `system.cpu.frequency` vs. `hw.cpu.speed`).

> Each metric needs a unique name. Here and below.

While for these I could use `.utilization`, `.rate` and `.usage`, and e.g. `throttled.time` is fine for total counter, something like `throttled.utilization` for ratio sounds completely bonkers.

For now I'll use throttling ratio without suffix, but that does not feel quite correct either.  Happily Prometheus plugin adds `_ratio` suffix to it, which at least makes it understandable to users.

Value, & value ratio (against max/limit) for frequency and temperature, are also problematic.  OTEL spec requires value itself not to have suffix, and as they are not consumable, but result of internal state & environmental factors [1], `.utilization` suffix is wrong for their ratios.

Btw. Several OTEL metrics specify just `.limit` metric without a ratio one.  While that is fine with frontends like Grafana / Prometheus, not having those (especially rate) when accessing collectd data directly (for validation & local GPU checks) sucks.  Switching from ratios to reporting limits would also be sad as Sysman plugin already has quite a bit of code for showing user specified mix of values, their rates and ratios for most metrics...

Suggestions?

----

[1] GPU load, cooling, air temperature

PS. OTEL fan metric actually use `_ratio` suffix, see: https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hwfan---fan-metrics (!)

(Sysman frontend provides API for Fan metrics, although backend support for that is still missing.)

octo replied:
The conventions for hardware are in the "experimental" status, i.e. now would be the time to propose changes. I'd encourage you to join the OpenTelemetry Slack and start a conversation there: https://slack.cncf.io/

Not every ratio needs to be a "utilization". For example, the "HTTP 5xx errors / total number of requests" ratio makes no sense expressed as "utilization". Likewise, the "overclocked / non-overclocked frequency" ratio is not a "utilization". IMHO `hw.gpu.frequency_ratio` is absolutely fine.

For temperature, I don't find a ratio meaningful. As a user, I'd prefer to get current value + limit metrics.

For throttled, I suggest we report a "throttled.time" metric that is scaled to seconds per second, and remove the ratio metric. I'd prefer to use the floating point counters for this once they're merged, but we can use a gauge for now.

eero-t replied:
Before using `_ratio` names, `write_prometheus` needs to be fixed to avoid duplicating that suffix => added that to list in PR description.

Yes, temperature ratio is probably redundant, as I would imagine GPUs to have quite similar limits.  For other metrics, ratios are nice as one does not need to code per-device rules for problematic situations, and for #4177 use-case a pre-calculated one is _much_ preferred.

Throttling time was in Sysman API from the start, but I've yet to seem backend support for it, so one option would be to ifdef it out, or split it to another branch, until there's backend support for it (frequency metric does have throttled attribute, but that's just a point sample).

PS. Just a reminder...  It will be 1+ week until I'll update this again (or review other PRs).

octo commented on src/gpu_sysman.c:
```diff
@@ -1259,20 +1265,26 @@ static bool gpu_mems_bw(gpu_device_t *gpu) {
+      .unit = "1",
   };
   metric_family_t fam_rate = {
       .help = "Memory bandwidth usage rate (in bytes per second)",
-      .name = METRIC_PREFIX "memory_bw_bytes_per_second",
+      .name = METRIC_PREFIX "memory.io.rate",
       .type = METRIC_TYPE_GAUGE,
+      .unit = "By/s",
   };
```
I'd drop this metric, we already report the same information as a counter.

Dito for `fabric.io.rate` below.

octo replied:
Many plugins have a `StoreRates` option that will translate counters to a rate. The same can easily be added for the *write_log plugin*.

eero-t replied:
> Many plugins have a `StoreRates` option that will translate counters to a rate.

So it seems:
```
$ git grep -l StoreRates src/write_*
src/write_graphite.c
src/write_http.c
src/write_influxdb_udp.c
src/write_kafka.c
src/write_mongodb.c
src/write_redis.c
src/write_riemann.c
src/write_sensu.c
src/write_syslog.c
src/write_tsdb.c
```

Unfortunately not `write_prometheus` which is really the only one I care about currently (in addition to Sysman plugin) does not support it.

> The same can easily be added for the _write_log plugin_.

`write_log` plugin is missing also other features making usable for local monitoring / debugging, discussed in #4177.

octo commented on src/gpu_sysman.c:
```diff
@@ -1644,13 +1658,15 @@ static bool gpu_freqs_throttle(gpu_device_t *gpu) {
   metric_family_t fam_ratio = {
       .help =
           "Ratio (0-1) of HW frequency being throttled during query interval",
-      .name = METRIC_PREFIX "throttled_ratio",
+      .name = METRIC_PREFIX "throttled",
       .type = METRIC_TYPE_GAUGE,
+      .unit = "1",
   };
   metric_family_t fam_counter = {
       .help = "Total time HW frequency has been throttled (in microseconds)",
-      .name = METRIC_PREFIX "throttled_usecs_total",
+      .name = METRIC_PREFIX "throttled.time",
       .type = METRIC_TYPE_COUNTER,
+      .unit = "us",
   };
```
Aren't these two metrics reporting the same thing, just scaled by 1,000,000?

eero-t replied:
They are for different time intervals / use-cases.  One is total time since (collectd) startup, which will be (soon) changed to be reported in seconds, and the other is ratio of time device was throttled between previous and current query.

Counter is more useful for things like Prometheus, ratio for direct monitoring.  I do not not expect both to be enabled at the same time.

octo replied:
As mentioned above, I think we should consider support for counters to be ubiquitous and not make design decisions based on debugging output.

eero-t replied:
This PR is about converting current Sysman plugin metric units & names to OTEL spec, not about changing plugin metric variant handling.  That discussion belonged to #4047 and #4173 (and maybe bit to #4177), not this one.

Maybe you could file a ticket where we can discuss what should be done about the variants?

eero-t replied:
Note that while metrics output directly from plugin is useful for debugging collectd, that's not its main use-case, but the local GPU monitoring (I'll add note of that in #4177).

eero-t commented on src/gpu_sysman.c:
```diff
@@ -1644,13 +1658,15 @@ static bool gpu_freqs_throttle(gpu_device_t *gpu) {
-      .name = METRIC_PREFIX "throttled_usecs_total",
+      .name = METRIC_PREFIX "throttled.time",
       .type = METRIC_TYPE_COUNTER,
+      .unit = "us",
```
Rebased to latest code, added commit changing all metrics to use base units, and updated PR description accordingly.

eero-t commented on src/daemon/resource.c:
```diff
@@ -51,7 +51,7 @@ static void otel_resource_attributes(void) {
   }
 
-  size_t tmp_sz = strlen(ra) + 1;
+  size_t tmp_sz = strlen(ra) + 2;
```
As its WIP commit states, its WA for GCC warning bug (on whole program analysis of the tester).  I need it while developing, but will drop it before merge.

eero-t commented on src/gpu_sysman.c:
```diff
@@ -810,12 +835,13 @@ static void ras_submit(gpu_device_t *gpu, const char *name, const char *help,
   };
   metric_t m = {0};
 
   m.value.counter = value;
```
Thanks, good catch! Fixed.

eero-t commented on src/gpu_sysman_test.c:
```diff
@@ -591,83 +614,104 @@ typedef struct {
+          "NOTE: '%s' value (%g) differs from expected (%g) by acceptable %g\n",
+          name, value, expected, diff);
+  return true;
+}
```
Ok, but it's a bit ugly, as it needs a wrapper (due to being macro that does `return`) & it doing unasked output.  `testing.h` inclusion needs WA too:
```
static int expect_double_eq(double expect, double actual) {
  /* WA for "unused-variable" warning on testing.h */
  fail_count__++;
  /* macro returns -1 on non-equality, continues if equal */
  EXPECT_EQ_DOUBLE(expect, actual);
  fail_count__--;
  return 0;
 }
```

eero-t commented:
Especially after #4291 is merged, this will be large enough that I'll split parts to a separate PR(s), most likely `write_prometheus` change, resource attributes and base units (FP counter) use, and few small fixes.

eero-t commented:
FYI: I started working on other Sysman plugin changes that I think should be in the release, before I get back to OTEL support (which definitely should be in release, so that metric names do not change after it).

Metric name changes require changing also other things than collectd (metric queries, validation tests, grafana charts), for which I haven't had time for, and such changes really should be done in one go, so that one does not need to support more than 2 version of metric names => I'd rather do OTEL related things as the last larger Sysman change before release.

eero-t commented:
Rebased on top of the latest plugin changes in `collectd-6.0` branch, fixed merge conflicts, and squashed fix commits (to commits they fix).

It's only build tested.   I'll try to split it into two separate PRs (next month), one with metric name changes, and another with rest of the changes, with name changes going in last (as that has impact also outside the plugin itself).

eero-t commented:
>  I'll try to split it into two separate PRs (next month), one with metric name changes, and another with rest of the changes, with name changes going in last (as that has impact also outside the plugin itself).

Closing this now that I split the changes to 3 separate PRs:
* https://github.com/collectd/collectd/pull/4299
* https://github.com/collectd/collectd/pull/4300
* https://github.com/collectd/collectd/pull/4301
